File: LeetCode/518.Coin_Change_II/Mario/518.Coin_Change_II.java
class Solution {
    public int change(int amount, int[] coins) {
        int m = coins.length;
        int[] dp = new int[amount + 1];

        for (int i = 0; i <= amount; i += coins[0]) {
            dp[i] = 1;
        }

        for (int i = 1; i < m; i++) {
            for (int j = coins[i]; j <= amount; j++) {
                dp[j] = dp[j] + dp[j - coins[i]];
            }
            // System.out.println(Arrays.toString(dp));
        }

        return dp[amount];
    }
}

File: LeetCode/518.Coin_Change_II/Mario/518.Coin_Change_II.py
#!/usr/bin/env python3


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        # dp[amount] means the num of combination
        # dp[i] = sum(dp[i-c]) for c in coins

        dp = [0] * (amount + 1)
        dp[0] = 1

        for c in coins:
            for i in range(c, amount + 1):
                dp[i] += dp[i - c]
                # print(dp)

        return dp[amount]
